src/InvariantTheory/invariant_rings.jl

```diff
@@ -0,0 +1,185 @@
+
+mutable struct InvRing{S, T, U, V, W, X}
+   field::S
+   poly_ring::T
+
+   group::U
+   action::Vector{V}          # Needs to be set (so far)
+   action_singular::Vector{W} # Needs to be set (so far)
+
+   modular::Bool
+
+   primary::Vector{X}
+   secondary::Vector{X}
+   irreducible_secondary::Vector{X}
+   fundamental::Vector{X}
+
+   # Cache some stuff on the Singular side
+   # (possibly removed at some point)
+   reynolds_singular::Singular.smatrix
+   molien_singular::Singular.smatrix
+   primary_singular::Singular.smatrix
+
+   function InvRing(K::S, G::U, action::Vector{V}) where {S, U, V}
+     n = degree(G)
```

Comment:
What is `G::U` supposed to be? You call `degree` on it... so perhaps a permutation group? Or is degree also supposed to be the dimension of the natural module of a matrix group?

Reply:
So far we can only handle matrix groups because we pass everything to Singular. So, `degree` is the rank of the matrices as defined in `Groups/matrices/MatGrp.jl:490`.
I would think that this constructor should not restrict the type or maybe `U <: AbstractAlgebra.Group` if absolutely necessary. The user is not supposed to call this anyway.